### PR TITLE
[enhance](function) refactor from_format_str and support more format

### PR DESCRIPTION
**Pull request by zclllyybb**
## Proposed changes

Issue Number: close #xxx

now support incomplete date parsing with any legal format, like:

```sql
mysql> select STR_TO_DATE('2020%2', '%Y%%%m');
+---------------------------------+
| str_to_date('2020%2', '%Y%%%m') |
+---------------------------------+
| 2020-02-01                      |
+---------------------------------+
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Review by Gabriel39 — DISMISSED** [review]

File: be/src/vec/runtime/vdatetime_value.cpp [hunk 1]
```diff
@@ -1199,26 +1191,41 @@ static int check_word(const char* lib[], const char* str, const char* end, const
...
-    bool date_part_used = false;
-    bool time_part_used = false;
-    bool frac_part_used = false;
-    bool already_set_time_part = false;
-
-    int day_part = 0;
+    bool already_set_time_part = false; // skip time part in the end's setting.
+
+    uint32_t part_used = 0;
+    constexpr int year_part = 1U << 0;
```
Comment (HappenLee) [inline be/src/vec/runtime/vdatetime_value.cpp:1205]: constexpr use YEAR_PART

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]

**Pull request merged**